File: server/app/models/forms/SirForm.scala
package models.forms

import java.util.UUID

import org.joda.time.DateTime
import slick.collection.heterogeneous.HNil
import slick.jdbc.SQLServerProfile.api._
import extensions.CustomColumnTypes._

case class SirForm (
                     id: Option[UUID],
                     sirNo: Option[String],
                     sirArea: Option[String],
                     playerId: Option[UUID],
                     sirMethodVerify: Option[String],
                     sirCurrentRelationship: Option[String],
                     sirAffiliationRelationship: Option[String],
                     sirTypeSuspiciousActivities: Option[String],
                     sirTranDateFrom: Option[DateTime],
                     sirTranDateTo: Option[DateTime],
                     ctrFormId: Option[UUID],
                     ctrFormNo: Option[String],
                     sirAmount: Option[Float],
                     sirSuspiciousIncidence: Option[String],
                     sirWatchlist: Option[Boolean],
                     sirStaffId: Option[String],
                     sirStaffDepartment: Option[String],
                     sirDocumentNo: Option[String],
                     submission: Option[Boolean],
                     status: Option[String],
                     recievedDate: Option[DateTime],
                     recievedBy: Option[String],
                     review: Option[Boolean],
                     sirAffiliationRelationshipOthers: Option[String],
                     sirCurrentRelationshipOthers: Option[String],
                     sirTypeSuspiciousActivitiesOthers: Option[String],
                     sirMethodVerifyOthers: Option[String],
                     isCloneToSTR: Option[Boolean],
                     sirTransactionType: Option[String],
                     playerWith: Option[String],
                     createdDate: Option[DateTime],
                     createdBy: Option[UUID],
                     modifiedDate: Option[DateTime],
                     modifiedBy: Option[UUID]
                   )


final class SirFormTable (tag: Tag) extends Table[SirForm](tag, Some("form"), "sirForm") {

  def id = column[Option[UUID]]("id", O.PrimaryKey)
  def sirNo = column[Option[String]]("sirNo")
  def sirArea = column[Option[String]]("sirArea")
  def playerId = column[Option[UUID]]("playerId")
  def sirMethodVerify = column[Option[String]]("sirMethodVerify")
  def sirCurrentRelationship = column[Option[String]]("sirCurrentRelationship")
  def sirAffiliationRelationship = column[Option[String]]("sirAffiliationRelationship")
  def sirTypeSuspiciousActivities = column[Option[String]]("sirTypeSuspiciousActivities")
  def sirTranDateFrom = column[Option[DateTime]]("sirTranDateFrom")
  def sirTranDateTo = column[Option[DateTime]]("sirTranDateTo")
  def ctrFormId = column[Option[UUID]]("ctrFormId")
  def ctrFormNo = column[Option[String]]("ctrFormNo")
  def sirAmount = column[Option[Float]]("sirAmount")
  def sirSuspiciousIncidence = column[Option[String]]("sirSuspiciousIncidence")
  def sirWatchlist = column[Option[Boolean]]("sirWatchlist")
  def sirStaffId = column[Option[String]]("sirStaffId")
  def sirStaffDepartment = column[Option[String]]("sirStaffDepartment")
  def sirDocumentNo = column[Option[String]]("sirDocumentNo")
  def submission = column[Option[Boolean]]("submission")
  def status = column[Option[String]]("status")
  def recievedDate = column[Option[DateTime]]("recievedDate")
  def recievedBy = column[Option[String]]("recievedBy")
  def review = column[Option[Boolean]]("review")
  def sirAffiliationRelationshipOthers = column[Option[String]]("sirAffiliationRelationshipOthers")
  def sirCurrentRelationshipOthers = column[Option[String]]("sirCurrentRelationshipOthers")
  def sirTypeSuspiciousActivitiesOthers = column[Option[String]]("sirTypeSuspiciousActivitiesOthers")
  def sirMethodVerifyOthers = column[Option[String]]("sirMethodVerifyOthers")
  def isCloneToSTR = column[Option[Boolean]]("isCloneToSTR")
  def sirTransactionType = column[Option[String]]("sirTransactionType")
  def playerWith = column[Option[String]]("playerWith")
  def createdDate = column[Option[DateTime]]("createdDate")
  def createdBy = column[Option[UUID]]("createdBy")
  def modifiedDate = column[Option[DateTime]]("modifiedDate")
  def modifiedBy = column[Option[UUID]]("modifiedBy")

  def * = (
    id ::
      sirNo ::
      sirArea ::
      playerId ::
      sirMethodVerify ::
      sirCurrentRelationship ::
      sirAffiliationRelationship ::
      sirTypeSuspiciousActivities ::
      sirTranDateFrom ::
      sirTranDateTo ::
      ctrFormId ::
      ctrFormNo ::
      sirAmount ::
      sirSuspiciousIncidence ::
      sirWatchlist ::
      sirStaffId ::
      sirStaffDepartment ::
      sirDocumentNo ::
      submission ::
      status ::
      recievedDate ::
      recievedBy ::
      review ::
      sirAffiliationRelationshipOthers ::
      sirCurrentRelationshipOthers ::
      sirTypeSuspiciousActivitiesOthers ::
      sirMethodVerifyOthers ::
      isCloneToSTR ::
      sirTransactionType ::
      playerWith ::
      createdDate ::
      createdBy ::
      modifiedDate ::
      modifiedBy :: HNil
    ).mapTo[SirForm]
}

File: client/app/view/forms/SIR/SIRForm.js
Ext.define("UI.view.forms.SIR.SIRForm", {
    extend: "UI.view.forms.AbstractForm",
    xtype: 'form-sir',
    requires: [
        'UI.view.form.SIR.SIRFormController',
        'UI.view.forms.SIR.SIRFormModel'
    ],
    controller: 'form-sir',
    scrollable: true,
    viewModel: 'form-sir',  
    initComponent: function() {
        var me = this;
        me.callParent(arguments);
    },
    initData: function() {
        var me = this,
        viewmodel = me.getViewModel(),
        state = me.extraData.state;

        viewmodel.set('formInfo', {
            playerWith: 'GERC Card Number'
        });

        switch(state) {
            case 'edit': {
                var data = me.extraData.record.getData();
                viewmodel.set('formInfo', data);
                viewmodel.set('readonly', false);    
                me.fireEvent('loadActivty', data.id);       
                me.fireEvent('loadAttachment', data.id);
                break;
            }
            case 'view': {
                var data = me.extraData.record.getData();
                viewmodel.set('formInfo', data);
                viewmodel.set('readonly', true);
                me.fireEvent('loadActivty', data.id);  
                me.fireEvent('loadAttachment', data.id);  
                me.isPreview = true;
            }
        }
    },
    buildInfoItems: function(parent) {
        return [{
            fieldLabel: 'SIR Form No',
            allowBlank: true,
            labelWidth: 190,
            emptyText: 'AUTO',
            editable: false,
            bind: {
                value: '{formInfo.sirNo}'
            }
        }, {
            fieldLabel: 'SIR Doc No',
            allowBlank: true,
            bind: {
                value: '{formInfo.sirDocumentNo}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'datefield',
            fieldLabel: 'Received Date',
            bind: {
                value: '{formInfo.recievedDate}',
                readOnly: '{readonly}'
            }
        }, {
            fieldLabel: 'Received By',
            labelWidth: 190,
            bind: {
                value: '{formInfo.recievedBy}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'x-component-lookup',
            key: 'SIR_AREA',
            fieldLabel: 'Area',
            bind: {
                value: '{formInfo.sirArea}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'combo',
            fieldLabel: 'Submission',
            allowBlank: true,
            valueField: 'value',
            displayField: 'name',
            value: false,
            store: {
                type: 'store',
                data: [
                    { value: true, name: 'Yes' },
                    { value: false, name: 'No' }
                ]
            },
            bind: {
                value: '{formInfo.submission}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'x-player-fieldset',
            title: 'Player Info',
            columnWidth: 1
        }, {
            xtype: 'x-component-lookup',
            key: 'SIR_METHOD',
            labelWidth: 250,
            fieldLabel: 'Method Used to Verify Identify',
            columnWidth: 0.5,
            bind: {
                value: '{formInfo.sirMethodVerify}',
                readOnly: '{readonly}'
            }
        }, {
            allowBlank: false,
            columnWidth: 0.5,
            hiddenLabel: 'Method Used to Verify Identify Others',
            hiddenAsterik: true,
            emptyText: 'Others',
            disabled: true,
            bind: {
                disabled: '{!(formInfo.sirMethodVerify == "Others")}',
                value: '{formInfo.sirMethodVerifyOthers}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'x-component-lookup',
            labelWidth: 250,
            fieldLabel: 'Current Relationship',
            key: 'SIR_CURRENT_RELATION',
            columnWidth: 0.5,
            bind: {
                value: '{formInfo.sirCurrentRelationship}',
                readOnly: '{readonly}'
            }
        }, {
            allowBlank: false,
            columnWidth: 0.5,
            hiddenLabel: 'Current Relationship Others',
            emptyText: 'Others',
            hiddenAsterik: true,
            disabled: true,
            bind: {
                disabled: '{!(formInfo.sirCurrentRelationship == "Others")}',
                value: '{formInfo.sirCurrentRelationshipOthers}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'x-component-lookup',
            labelWidth: 250,
            fieldLabel: 'Affiliation or Relationship to NagaWorld',
            columnWidth: 0.5,
            key: 'SIR_AFFILIATION',
            columnWidth: 0.5,
            bind: {
                value: '{formInfo.sirAffiliationRelationship}',
                readOnly: '{readonly}'
            }
        }, {
            allowBlank: false,
            hiddenLabel: 'Affiliation or Relationship to NagaWorld Others',
            emptyText: 'Others',
            hiddenAsterik: true,
            columnWidth: 0.5,
            disabled: true,
            bind: {
                disabled: '{!(formInfo.sirAffiliationRelationship == "Others")}',
                value: '{formInfo.sirAffiliationRelationshipOthers}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'x-component-lookup',
            labelWidth: 250,
            fieldLabel: 'Type of Suspicious Activities',
            columnWidth: 0.5,
            key: 'SIR_TYPE_SUSPICIOUS',
            bind: {
                value: '{formInfo.sirTypeSuspiciousActivities}',
                readOnly: '{readonly}'
            }
        }, {
            allowBlank: false,
            columnWidth: 0.5,
            disabled: true,
            emptyText: 'Others',
            hiddenLabel: 'Type of Suspicious Activities Others',
            hiddenAsterik: true,
            bind: {
                disabled: '{!(formInfo.sirTypeSuspiciousActivities == "Others")}',
                value: '{formInfo.sirTypeSuspiciousActivitiesOthers}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'x-component-lookup',
            labelWidth: 250,
            fieldLabel: 'Transaction Type',
            columnWidth: 1,
            key: 'SIR_TRANSACTION_TYPE',
            bind: {
                value: '{formInfo.sirTransactionType}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'datetime',
            labelWidth: 250,
            columnWidth: 0.5,
            fieldLabel: 'From Trans Date/Time',
            bind: {
                value: '{formInfo.sirTranDateFrom}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'datetime',
            labelWidth: 150,
            columnWidth: 0.5,
            fieldLabel: 'To Trans Date/Time',
            bind: {
                value: '{formInfo.sirTranDateTo}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'ctrpickup',
            fieldLabel: 'CTR Form No',
            allowBlank: true,
            labelWidth: 250,
            columnWidth: 0.5,
            bind: {
                value: '{formInfo.ctrFormId}',
                selection: '{ctrFormSelection}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'currencyfield',
            fieldLabel: 'Amount',
            value: 0.00,
            columnWidth: 0.5,
            labelWidth: 150,
            bind: {
                value: '{formInfo.sirAmount}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'label',
            text: 'Suspicious Incidence Report Narrative:',
            columnWidth: 1,
            margin: '0 0 10 0'
        }, {
            xtype: 'textarea',
            maxLength: 2000,
            columnWidth: 1,
            allowBlank: true,
            height: 120,
            bind: {
                value: '{formInfo.sirSuspiciousIncidence}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'combo',
            labelWidth: 355,
            columnWidth: 0.5,
            fieldLabel: 'Name search checked against internal database watchlist?',
            value: false,
            valueField: 'value',
            displayField: 'name',
            store: {
                type: 'store',
                data: [
                    { value: true, name: 'Yes' },
                    { value: false, name: 'No' }
                ]
            },
            bind: {
                value: '{formInfo.sirWatchlist}',
                readOnly: '{readonly}'
            }
        }, {
            fieldLabel: 'Staff ID',
            columnWidth: 0.2,
            labelWidth: 70,
            bind: {
                value: '{formInfo.sirStaffId}',
                readOnly: '{readonly}'
            }
        }, {
            xtype: 'x-component-lookup',
            columnWidth: 0.3,
            labelWidth: 115,
            key: 'AML_DEPARTMENT',
            fieldLabel: 'Staff Department',
            bind: {
                value: '{formInfo.sirStaffDepartment}',
                readOnly: '{readonly}'
            }
        },];
    },
    getFormConfiguration: function(form) {
        var me = this,
        extraData = me.extraData,
        state = extraData.state;

        if(state == 'new') {
            config = {
                commentBlock: {
                    visible: false
                },
                formAttachmentBlock: {
                    visible: true,
                    editable: true
                },
                activityBlock: {
                    visible: false
                },
                exportPdf: {
                    visible: false
                }
            }
        } else if (state == 'edit') {
            config = {
                commentBlock: {
                    visible: true
                },
                formAttachmentBlock: {
                    visible: true,
                    editable: true
                },
                activityBlock: {
                    visible: true
                },
                exportPdf: {
                    visible: true
                }
            };
        } else {
            config = {
                commentBlock: {
                    visible: false
                },
                formAttachmentBlock: {
                    visible: true,
                    editable: false
                },
                activityBlock: {
                    visible: true
                },
                exportPdf: {
                    visible: true
                }
            };
        }

        return config;
    }
});